File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class Main {

    static int N, cnt;
    static boolean visited[][];
    static char array[][];
    static int arrX[] = {-1, 0, 1, 0};
    static int arrY[] = {0, 1, 0, -1};

    public static void main(String[] args) throws IOException {

        Scanner scanner = new Scanner(System.in);

        N = scanner.nextInt();

        array = new char[N][N];
        visited = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            array[i] = scanner.next().toCharArray();
        }


        int number = 1;
        ArrayList<Integer> answer = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (array[i][j] == '1') {
                    cnt = 0;
                    dfs(i, j, number);
                    answer.add(cnt + 1);
                    number++;
                }

            }
        }


        System.out.println(number - 1);

        Collections.sort(answer);
        for (int i = 0; i < answer.size(); i++) {
            System.out.println(answer.get(i));
        }

    }

    public static void dfs(int x, int y, int num) {
        if (visited[x][y]) {
            return;
        }

        visited[x][y] = true;

        for (int i = 0; i < 4; i++) {
            int defX = x + arrX[i];
            int defY = y + arrY[i];

            if (defX >= 0 && defX < N && defY >= 0 && defY < N && !visited[defX][defY] && array[defX][defY] == '1') {
                array[defX][defY] = (char) num;
                cnt++;
                dfs(defX, defY, num);
            }
        }

    }

}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
# 2667 단지번호붙이기 [DFS]

from collections import deque

# 입력 받기
N = int(input())
grid = [list(input()) for _ in range(N)]
visited = [[False] * N for _ in range(N)]

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

def dfs(x, y):
    stack = deque([(x, y)])
    visited[x][y] = True
    count = 1

    while stack:
        x, y = stack.pop()
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < N and 0 <= ny < N and not visited[nx][ny] and grid[nx][ny] == '1':
                stack.append((nx, ny))
                visited[nx][ny] = True
                count += 1
    return count

# 각 구역의 크기를 저장할 리스트
sizes = []

# 모든 칸에 대해 DFS 실행
for i in range(N):
    for j in range(N):
        if grid[i][j] == '1' and not visited[i][j]:
            size = dfs(i, j)
            sizes.append(size)


print(len(sizes)) 
sizes.sort()
for size in sizes:
    print(size)  